controllers/ingressnodefirewall_controller_rules_test.go

```diff
@@ -794,14 +610,14 @@ var _ = Describe("IngressNodeFirewall controller rules", func() {
 					Expect(k8sClient.Create(ctx, &ingressNodeFirewall)).Should(Succeed())
 				}
 
-				By("Waiting for the expected list of IngressNodeFirewallNodeStates")
-				hasIngressNodeFirewallNodeStates(ctx, k8sClient, []string{worker0Name})
```

Comment:
Ahhh ok I think I get this now. So the behavior changed and not the node state isn't generated any more in case of issues and instead is gone?  Still, can't we rewrite the above tests instead of dropping them? Otherwise, we always only test for success

Reply:
we need to enable ut CI lane so will be good if we have stable run then I find away to add failing condition some to this test suite in different PR if that is ok ?

Reply:
I just checked the unit tests, and they uncover issues with the current code that weren't there before:
~~~
1.6613392710556202e+09	ERROR	controllers.IngressNodeFirewall	Failed to create new IngressNodeFirewallNodeState	{"req.Name": "firewall-0", "ingressNodeFirewallNodeState.Namespace": "ingress-node-fw-config-test-namespace", "ingressNodeFirewallNodeState.Name": "worker-0", "error": "IngressNodeFirewallNodeState.ingressnodefirewall.openshift.io \"worker-0\" is invalid: spec.interfaceIngressRules.eth0: Invalid value: \"null\": spec.interfaceIngressRules.eth0 in body must be of type array: \"null\""}
~~~

That's logged by the reconciler. So some previous update broke the reconciler. I'm really reluctant to remove unit tests that expose valid issues just to make them pass ;-)

Reply:
This isn't the correct location for this, but with this workaround here, 3 out of the 4 failing unit tests pass:
~~~
--- a/controllers/ingressnodefirewall_controller.go
+++ b/controllers/ingressnodefirewall_controller.go
@@ -168,6 +168,13 @@ func (r *IngressNodeFirewallReconciler) Reconcile(ctx context.Context, req ctrl.
                        },
                        Spec: desiredNodeState.Spec,
                }
+               if ingressNodeFirewallNodeState.Spec.InterfaceIngressRules != nil {
+                       for k := range ingressNodeFirewallNodeState.Spec.InterfaceIngressRules {
+                               if ingressNodeFirewallNodeState.Spec.InterfaceIngressRules[k] == nil {
+                                       ingressNodeFirewallNodeState.Spec.InterfaceIngressRules[k] = []infv1alpha1.IngressNodeFirewallRules{}
+                               }
+                       }
+               }
                err = r.Create(ctx, &ingressNodeFirewallNodeState)
                if err != nil {
                        r.Log.Error(err, "Failed to create new IngressNodeFirewallNodeState",
~~~

So on merge, make sure that [eth0] or any interface key is never nil but that it's a map

Reply:
lemme check 30 mins to see where that really has to go

Reply:
Here's the fix, don't change any of the unit tests, instead just push this fix:
~~~
--- a/controllers/ingressnodefirewall_controller.go
+++ b/controllers/ingressnodefirewall_controller.go
@@ -378,7 +378,7 @@ func mergeRuleSet(a, b []infv1alpha1.IngressNodeFirewallRules) ([]infv1alpha1.In
                        // Now, go over each existing rule in the already merged slice.
                        for i, ruleA := range a {
                                if len(ruleA.SourceCIDRs) != 1 {
-                                       return nil, fmt.Errorf(
+                                       return []infv1alpha1.IngressNodeFirewallRules{}, fmt.Errorf(
                                                "cannot merge into ruleset A with invalid SourceCIDRs: '%v'", ruleA.SourceCIDRs)
                                }
                                // If the CIDR already exists in A, then merge it in and continue with the next CIDR.
@@ -386,7 +386,7 @@ func mergeRuleSet(a, b []infv1alpha1.IngressNodeFirewallRules) ([]infv1alpha1.In
                                        a[i].FirewallProtocolRules, err = mergeFirewallProtocolRules(
                                                ruleA.FirewallProtocolRules, ruleB.FirewallProtocolRules)
                                        if err != nil {
-                                               return nil, err
+                                               return []infv1alpha1.IngressNodeFirewallRules{}, err
                                        }
                                        continue withNextSourceCIDR
                                }
~~~